### DISCUSSION TITLE
How to use plugins when using kubernetes

### DISCUSSION THREAD

**Discussion by Officeyutong** [WasmEdge/WasmEdge#2327] [Q&A] [answered]
I'm following the the instructions at https://wasmedge.org/book/en/use_cases/kubernetes/quickstart.html to set up `CRI-O and crun`.

After installation, I can successfully run the demo image through `wget -qO- https://raw.githubusercontent.com/second-state/wasmedge-containers-examples/main/kubernetes_crio/simple_wasi_application.sh | bash
`.

But when I try to run an wasm image which uses a module `wasm_bpf` provided by a plugin I wrote, the following error came out:
```console
root@newoj:~/test-docker# cat run.sh
../crio-install/kubernetes/cluster/kubectl.sh run -it --rm --restart=Never opensnoop --image docker.io/officeyutong/opensnoop-wasm:latest --annotations="module.wasm.image/variant=compat-smart" /opensnoop.wasm
root@newoj:~/test-docker# ./run.sh
[2023-03-01 08:09:06.657] [error] instantiation failed: unknown import, Code: 0x62
[2023-03-01 08:09:06.657] [error]     When linking module: "wasm_bpf" , function name: "wasm_load_bpf_object"
[2023-03-01 08:09:06.657] [error]     At AST node: import description
[2023-03-01 08:09:06.657] [error]     At AST node: import section
[2023-03-01 08:09:06.657] [error]     At AST node: module
[libcrun:wasmedge]: could not get wasmedge result from VM
pod "opensnoop" deleted
pod default/opensnoop terminated (Error)
```

It seems that the plugin wasn't loaded. But I've already put the plugin under proper directories and set `WASMEDGE_PLUGIN_PATH` to the corresponding value.

```console
root@newoj:~/test-docker# ls /usr/local/lib/wasmedge/
libwasmedgePluginWasmBpf.so  libwasmedgePluginWasmEdgeProcess.so
root@newoj:~/test-docker# echo $WASMEDGE_PLUGIN_PATH
/usr/local/lib/wasmedge/
root@newoj:~/test-docker# ls ~/.wasmedge/plugin/
libwasmedgePluginWasmBpf.so
```

But if I directly use the installed `wasmedge` to run the Wasm file, it will come successfully.

```console
root@newoj:~/test-docker# /usr/local/bin/wasmedge opensnoop.wasm

attach ok
PID    COMM              FD ERR PATH
4061354 sleep              3   0 /etc/ld.so.cache
4061354 sleep              3   0 /lib/x86_64-linux-gnu/libc.so.6
4061354 sleep              3   0 /usr/lib/locale/locale-archive
635    redis-server      22   0 /proc/635/stat
```

So my question is, how can I let plugins properly loaded when run a container through k8s?

**Comment by hydai** [comment, marked as answer]
Hi @Officeyutong 
Thanks for raising this. We don't allow loading plugins from the crun integration to avoid potential security issues currently. To enable this, you should call [an extra function](https://github.com/WasmEdge/WasmEdge/blob/master/include/api/wasmedge/wasmedge.h#L3385-L3394) in `libcrun/handler/wasmedge`. So you need to modify the crun source code and rebuild it.

We are also thinking about how to integrate the plugins more securely. If you have any ideas, please let me know.

**Reply by Officeyutong** [reply]
Thank you!
